webextension/privileged/ui/CloudDownloadsView.jsm

```diff
@@ -261,6 +261,15 @@ var CloudDownloadsView = {
         return;
       }
 
+      // Exit toggleAPIEnabledState if we don't have a valid styles URL yet
+      // This is seen on restart of browser after add-on install when
+      // CloudDownloadsView load triggers toggleAPIEnabledState (Fixes issue #51)
```

Comment:
You're describing only part of what happened, so I'm still confused. Why does CloudDownloadsView get loaded (if not from the init method you reference)? Why doesn't this happen every time the browser is started? Can't we make CloudDownloadsView not get loaded if we're not ready for that yet (ie fix the callsite) ?

Even in #51 I don't really understand what happens - is this an in-process restart? Or just shutting down the browser, and then this happens the next time you start the browser? Why does this only affect the 1st restart after installing the add-on, but not any of the other ones, as this code comment implies? The comments in #51 so far don't mention that this is a one-off problem...

Ultimately this feels like a wallpaper fix, rather than addressing the root cause.

Reply:
@gijsk 
The way API is designed - toggleAPIEnabledState can be triggered in three ways 
a) https://github.com/mozilla/shield-cloudstorage/blob/master/webextension/privileged/ui/CloudDownloadsView.jsm#L857

when CloudDownloadsView module is first imported


b)  After load -  cloud.services.api.enabled state toggles 
https://github.com/mozilla/shield-cloudstorage/blob/master/webextension/privileged/ui/CloudDownloadsView.jsm#L843

this is used by uninit and init method in cloudstorage.js here:
https://github.com/mozilla/shield-cloudstorage/blob/master/webextension/privileged/ui/cloudstorage.js#L56
https://github.com/mozilla/shield-cloudstorage/blob/master/webextension/privileged/ui/cloudstorage.js#L64


c) inside init method on subsequent browser restart after add-on install when api enabled is true 

https://github.com/mozilla/shield-cloudstorage/blob/master/webextension/privileged/ui/cloudstorage.js#L50

b) and c)  is triggered by cloudstorage.js with stylesURL property set inside init method.

It’s toggleAPIEnabledState triggered by a) on import here https://github.com/mozilla/shield-cloudstorage/blob/master/webextension/privileged/ui/cloudstorage.js#L23

that should not continue if we don’t have a valid stylesURL property set.  Alternate fix is to remove this call but my understanding we need this call to support initializing API.

Reply:
This issue happens for every restart of browser after add-on install. Add-on install sets cloud.services.api.enabled to true ( using init method mentioned in flow b))

Reply:
> that should not continue if we don’t have a valid stylesURL property set. Alternate fix is to remove this call but my understanding we need this call to support initializing API.

What does this mean, "to support initializing API" ?

Right now, even the code that sets the `stylesURL` property will first import CloudDownloadsView.jsm to actually get a reference to that object, which will already call this method, even before stylesURL is set, right? So I think calling toggleAPIEnabled at the end of the jsm is probably just wrong, if toggling it really requires `stylesURL` to be present.

That said, maybe we should also keep the early return if it's possible for the pref to be changed before anyone has set the stylesURL property. Though I guess that won't happen with the add-on as written...

Reply:
Perhaps it would make sense to have an explicit init method that takes a stylesURL argument, and ensure we invoke that explicitly from the webextension API consumer, which anyway already has an init method. Ditto for the other initialization things we're currently setting directly on the CloudDownloadsView singleton.

Reply:
@gijsk By initializing API I mean jsm has loaded and listening to pref changes. If we remove toggleAPIEnabled at the end of the jsm, toggleAPIEnabled is not getting invoked first time when add-on is installed.

It seems when add-on is installed getAPI call in Web Extension is executed after the init method in privileged/ui/cloudstorage.js 

so sequence of call is first init 
https://github.com/mozilla/shield-cloudstorage/blob/master/webextension/privileged/ui/cloudstorage.js#L56

and then getAPI that loads jsm module
https://github.com/mozilla/shield-cloudstorage/blob/master/webextension/privileged/ui/cloudstorage.js#L23,

Reply:
Is it possible that code that lazy load when we are setting stylesURL property hasn't finished loading before the pref is set to true here https://github.com/mozilla/shield-cloudstorage/blob/master/webextension/privileged/ui/cloudstorage.js#L56

and needs toggleAPIEnabled at the end of JSM module?

Reply:
I think I get it, once we remove toggleAPIEnabledState at the end of jsm module, we are relying on lazypreferencegetter here to set listeners and call toggleAPIEnabledState when api.enabled pref changes

https://github.com/mozilla/shield-cloudstorage/blob/master/webextension/privileged/ui/CloudDownloadsView.jsm#L843

But these listeners are not set until first gIsAPIEnabled is needed, which is used first inside toggleAPIEnabled, is that true?

Reply:
It sounds like we could fix this bug by:

- removing the direct `toggleAPIEnabledState()` call at the end of the jsm
- calling `toggleAPIEnabledState()` after setting the two prefs in https://github.com/mozilla/shield-cloudstorage/blob/master/webextension/privileged/ui/cloudstorage.js#L52-L57 . So basically, instead of having an if/else, you could just only set the prefs to true if the pref isn't true, and then always call `await CDV.toggleAPIEnabledState()`.

That would work, right?

Reply:
@gijsk it should work except  in one flow when user manually disable and re-enable add-on from about:addons. 

It seems even after add-on is disabled, CloudDownloadView jsm is sill listening to api.enabled pref flip here
https://github.com/mozilla/shield-cloudstorage/blob/master/webextension/privileged/ui/CloudDownloadsView.jsm#L843

and when add-on is renabled it calls toggleAPIEnabledState  along with toggleAPIEnabledState call in https://github.com/mozilla/shield-cloudstorage/blob/master/webextension/privileged/ui/cloudstorage.js#L43 resulting in duplicate ping. 

Is there a way to unset listener on lazypreference getter?

Reply:
Use `Cu.unload()` to unload CloudDownloadsView.jsm when the add-on is disabled?